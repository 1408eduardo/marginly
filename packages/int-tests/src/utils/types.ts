import { Wallet } from 'ethers';
import { FiatTokenV2_1Contract } from '../contract-api/FiatTokenV2';
import { MarginlyPoolContract } from '../contract-api/MarginlyPool';
import { NonfungiblePositionManagerContract } from '../contract-api/NonfungiblePositionManager';
import { SwapRouterContract } from '../contract-api/SwapRouter';
import { UniswapV3PoolContract } from '../contract-api/UniswapV3Pool';
import { WETH9Contract } from '../contract-api/WETH9';

export type ContractsCollection = {
  weth: WETH9Contract;
  usdc: FiatTokenV2_1Contract;
  uniswap: UniswapV3PoolContract;
  marginly: MarginlyPoolContract;
  nonFungiblePositionManager: NonfungiblePositionManagerContract;
  swapRouter: SwapRouterContract;
};
export type Suite = (contracts: ContractsCollection, treasury: Wallet, accounts: Wallet[]) => Promise<void>;
