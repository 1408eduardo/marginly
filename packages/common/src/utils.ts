import * as ethers from 'ethers';
import { BigNumber } from '@ethersproject/bignumber';

export function formatBalance(amount: BigNumber, decimals: number, assetSymbol: string): string {
  return `${amount} (${ethers.utils.formatUnits(amount, decimals)} ${assetSymbol.toUpperCase()})`;
}

export const sleep = (ms: number) => {
  return new Promise((resolve) => setTimeout(resolve, ms));
};

export const waitForTx = async (provider: ethers.providers.Provider, hash: string) => {
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const receipt = await provider.getTransactionReceipt(hash);
    if (!receipt) {
      await sleep(3000);
    } else {
      if (!receipt.status) {
        throw new Error('Transaction failed');
      }
      break;
    }
  }
};
