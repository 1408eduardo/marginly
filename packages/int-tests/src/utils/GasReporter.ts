import { ContractReceipt, ContractTransaction } from 'ethers';
import { logger } from './logger';
import { promises as fsPromises, existsSync } from 'fs';
import { join } from 'path';

type GasUsage = {
  max: number;
  min: number;
  avg: number;
  count: number;
};

export class GasReporter {
  private gasUsageStatistics: { [key: string]: GasUsage } = {};
  private gasUsage: { [key: string]: [number, number][] } = {};

  constructor(private suiteName: string) {}

  public async saveGasUsage(
    txName: string,
    x: ContractReceipt | Promise<ContractReceipt> | ContractTransaction | Promise<ContractTransaction>
  ) {
    const resolved = await x;
    let txReceipt: ContractReceipt;
    if ('wait' in resolved) {
      txReceipt = await resolved.wait();
    } else if ('gasUsed' in resolved) {
      txReceipt = resolved;
    } else {
      throw Error('Uknown argument');
    }

    const gasUsed = txReceipt.gasUsed.toNumber();
    const blockNumber = txReceipt.blockNumber;

    logger.debug(`â›½ Gas used: ${txName}    ${gasUsed}`);

    const existedStatistic = this.gasUsageStatistics[txName];
    if (existedStatistic) {
      existedStatistic.max = existedStatistic.max < gasUsed ? gasUsed : existedStatistic.max;
      existedStatistic.min = existedStatistic.min > gasUsed ? gasUsed : existedStatistic.min;
      existedStatistic.avg = Math.floor(
        (existedStatistic.avg * existedStatistic.count + gasUsed) / (existedStatistic.count + 1)
      );
      existedStatistic.count++;
    } else {
      this.gasUsageStatistics[txName] = {
        max: gasUsed,
        min: gasUsed,
        avg: gasUsed,
        count: 1,
      };
    }

    if (!this.gasUsage[txName]) {
      this.gasUsage[txName] = [[blockNumber, gasUsed]];
    } else {
      this.gasUsage[txName].push([blockNumber, gasUsed]);
    }

    return txReceipt;
  }

  public reportToConsole() {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const gasUsageStatistics = this.gasUsageStatistics;
    if (Object.keys(gasUsageStatistics).length > 0) {
      setTimeout(function () {
        console.log('Gas usage statistics');
        console.table(gasUsageStatistics);
      }, 10);
    }
  }

  async saveToFile() {
    const dir = join(__dirname, '../../', '__gas-usage__');

    if (!existsSync(dir)) {
      await fsPromises.mkdir(dir, { recursive: true });
    }

    for (const txName of Object.keys(this.gasUsage)) {
      const data = [`blockNumber,gasUsed\n`];
      for (const [blockNumber, gasUsed] of this.gasUsage[txName]) {
        data.push(`${blockNumber},${gasUsed}\n`);
      }
      const filename = `${this.suiteName}.${txName}.gas.csv`;

      await fsPromises.writeFile(join(dir, filename), data, {
        flag: 'w',
      });
    }
  }
}
