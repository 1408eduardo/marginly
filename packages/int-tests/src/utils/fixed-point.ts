import { BigNumber } from 'ethers';
import bn from 'bignumber.js';

export const FP96 = {
  one: BigNumber.from(2n ** 96n),
};

export const FP48 = {
  one: BigNumber.from(2n ** 48n),
};

// 365.25 * 24 * 60 * 60 * 2 ** 96
export const secondsInYearX96 = BigNumber.from('2500250661360148260042022567123353600');

export function pow(self: BigNumber, exponent: number): BigNumber {
  let result = FP96.one;
  while (exponent > 0) {
    if ((exponent & 1) == 1) {
      result = result.mul(self).div(FP96.one);
    }
    self = self.mul(self).div(FP96.one);
    exponent = exponent >> 1;
  }

  return result;
}

export function powTaylor(self: BigNumber, exponent: number): BigNumber {
  const x = self.sub(FP96.one);
  if (x >= FP96.one) {
    throw new Error(`x can't be greater than FP.one, series diverges`);
  }

  let resultX96 = FP96.one;
  let multiplier: BigNumber;
  let term = FP96.one;

  const steps = exponent < 3 ? exponent : 3;
  for (let i = 0; i != steps; ++i) {
    multiplier = BigNumber.from(exponent - i)
      .mul(x)
      .div(BigNumber.from(i + 1));
    term = term.mul(multiplier).div(FP96.one);
    resultX96 = resultX96.add(term);
  }

  return resultX96;
}

export function toHumanString(fp96Value: BigNumber): string {
  return bn(fp96Value.toString()).div(FP96.one.toString()).toString();
}

export function fp48ToHumanString(fp10Value: BigNumber): string {
  return bn(fp10Value.toString()).div(FP48.one.toString()).toString();
}
