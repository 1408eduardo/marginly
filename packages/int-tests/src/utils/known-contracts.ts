import { BaseContract, BigNumberish, providers, Signer } from 'ethers';
import WETH9 from '../contract-api/WETH9';
import FiatTokenV2_1 from '../contract-api/FiatTokenV2';
import UniswapV3Factory from '../contract-api/UniswapV3Factory';
import UniswapV3Pool from '../contract-api/UniswapV3Pool';
import SwapRouter from '../contract-api/SwapRouter';
import NonfungiblePositionManager from '../contract-api/NonfungiblePositionManager';

export const wethContract = (signerOrProvider?: Signer | providers.Provider) =>
  WETH9.connect(`0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2`, signerOrProvider);

export const usdcContract = (signerOrProvider?: Signer | providers.Provider) =>
  FiatTokenV2_1.connect(`0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48`, signerOrProvider);

export const uniswapFactoryContract = (signerOrProvider?: Signer | providers.Provider) =>
  UniswapV3Factory.connect(`0x1F98431c8aD98523631AE4a59f267346ea31F984`, signerOrProvider);

export const uniswapPoolContract = UniswapV3Pool.connect;

export const swapRouterContract = (signerOrProvider?: Signer | providers.Provider) =>
  SwapRouter.connect('0xE592427A0AEce92De3Edee1F18E0157C05861564', signerOrProvider);

export const nonFungiblePositionManagerContract = (signerOrProvider?: Signer | providers.Provider) =>
  NonfungiblePositionManager.connect('0xC36442b4a4522E871399CD717aBDD847Ab11FE88', signerOrProvider);

// Uniswap V3: Positions NFT 0xC36442b4a4522E871399CD717aBDD847Ab11FE88
