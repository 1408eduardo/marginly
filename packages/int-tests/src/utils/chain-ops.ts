import { logger } from '../utils/logger';
import { SECS_PER_BLOCK } from './const';
import { Web3Provider } from '@ethersproject/providers';
import { MarginlyPoolContract } from '../contract-api/MarginlyPool';
import { BigNumber, ethers, ContractReceipt } from 'ethers';

export async function waitBlocks(blocks: number): Promise<void> {
  logger.info(`Waiting for ${blocks} blocks`);
  return await new Promise((rs) => setTimeout(rs, blocks * SECS_PER_BLOCK * 1000.0));
}

export class Web3ProviderDecorator {
  readonly provider: Web3Provider;

  constructor(provider: Web3Provider) {
    this.provider = provider;
  }

  mineAtTimestamp(timestampSeconds: number): Promise<any> {
    return this.provider.send('evm_mine', [timestampSeconds]);
  }

  async getLastBlockTimestamp(): Promise<number> {
    return (await this.provider.getBlock(this.provider._lastBlockNumber)).timestamp;
  }
}

export type SwapEvent = {
  amount0: BigNumber;
  amount1: BigNumber;
  sqrtPriceX96: BigNumber;
  liquidity: BigNumber;
  tick: number;
};

export function decodeSwapEvent(txReceipt: ContractReceipt, uniswapAddress: string): SwapEvent {
  const swapEvent = txReceipt.events!.find((e) => e.address == uniswapAddress);
  const swapEventTypes = ['int256', 'int256', 'uint160', 'uint128', 'int24'];
  const result = ethers.utils.defaultAbiCoder.decode(swapEventTypes, swapEvent!.data);

  return {
    amount0: BigNumber.from(result[0]),
    amount1: BigNumber.from(result[1]),
    sqrtPriceX96: BigNumber.from(result[2]),
    liquidity: BigNumber.from(result[3]),
    tick: result[4],
  };
}

export async function getLongSortKeyX48(
  marginlyPool: MarginlyPoolContract,
  accountAddress: string
): Promise<BigNumber> {
  const position = await marginlyPool.positions(accountAddress);
  const index = BigNumber.from(position.heapPosition).sub(1);
  logger.debug(`  heap position is ${position.heapPosition}`);
  const [, leverage] = await marginlyPool.getLongHeapPosition(index);
  return BigNumber.from(leverage.key);
}

export async function getShortSortKeyX48(
  marginlyPool: MarginlyPoolContract,
  accountAddress: string
): Promise<BigNumber> {
  const position = await marginlyPool.positions(accountAddress);
  const index = BigNumber.from(position.heapPosition).sub(1);
  logger.debug(`  heap position is ${position.heapPosition}`);
  const [, leverage] = await marginlyPool.getShortHeapPosition(index);
  return BigNumber.from(leverage.key);
}
