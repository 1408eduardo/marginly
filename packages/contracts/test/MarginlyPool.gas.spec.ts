import { createMarginlyPool, getInitializedPool } from './shared/fixtures';
import { loadFixture, time } from '@nomicfoundation/hardhat-network-helpers';
import { ethers } from 'hardhat';
import snapshotGasCost from '@uniswap/snapshot-gas-cost';
import { BigNumber } from 'ethers';
import { toHumanString } from './shared/utils';
import { expect } from 'chai';

describe('Open position:', () => {
  it('depositBase', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, signer] = await ethers.getSigners();
    const depositAmount = 1000;

    await snapshotGasCost(await marginlyPool.connect(signer).depositBase(depositAmount));
  });

  it('depositQuote', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, signer] = await ethers.getSigners();
    const depositAmount = 1000;

    await snapshotGasCost(await marginlyPool.connect(signer).depositQuote(depositAmount));
  });
});

describe('Deposit into existing position:', () => {
  it('depositBase', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, signer1] = await ethers.getSigners();
    const firstDeposit = 2468;
    const secondDeposit = 2837;

    await marginlyPool.connect(signer1).depositBase(firstDeposit);
    await snapshotGasCost(marginlyPool.connect(signer1).depositBase(secondDeposit));
  });

  it('depositQuote', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, signer1] = await ethers.getSigners();
    const firstDeposit = 2468;
    const secondDeposit = 2837;

    await marginlyPool.connect(signer1).depositQuote(firstDeposit);
    await snapshotGasCost(marginlyPool.connect(signer1).depositQuote(secondDeposit));
  });
});

describe('System initialized:', async () => {
  it('long', async () => {
    const { marginlyPool, wallets } = await loadFixture(getInitializedPool);
    const longer = wallets[0];
    await marginlyPool.connect(longer).depositBase(1000);
    await snapshotGasCost(marginlyPool.connect(longer).long(900));
  });

  it('short', async () => {
    const { marginlyPool, wallets } = await loadFixture(getInitializedPool);
    const shorter = wallets[0];
    await marginlyPool.connect(shorter).depositQuote(3000);
    await snapshotGasCost(marginlyPool.connect(shorter).short(400));
  });

  it('depositBase', async () => {
    const { marginlyPool, wallets } = await loadFixture(getInitializedPool);
    const lender = wallets[0];
    await snapshotGasCost(marginlyPool.connect(lender).depositBase(100));
  });

  it('depositQuote', async () => {
    const { marginlyPool, wallets } = await loadFixture(getInitializedPool);
    const lender = wallets[0];
    await snapshotGasCost(marginlyPool.connect(lender).depositQuote(3000));
  });

  it('closePosition', async () => {
    const { marginlyPool } = await loadFixture(getInitializedPool);
    const signers = await ethers.getSigners();
    const borrower = signers[signers.length - 1];
    await snapshotGasCost(marginlyPool.connect(borrower).closePosition());
  });

  it('withdrawBase', async () => {
    const { marginlyPool } = await loadFixture(getInitializedPool);
    const signers = await ethers.getSigners();
    const longer = signers[signers.length - 1];
    await snapshotGasCost(marginlyPool.connect(longer).withdrawBase(100));
  });

  it('withdrawQuote', async () => {
    const { marginlyPool } = await loadFixture(getInitializedPool);
    const signers = await ethers.getSigners();
    const shorter = signers[11];
    await snapshotGasCost(marginlyPool.connect(shorter).withdrawQuote(100));
  });
});

describe('mc happens:', async () => {
  it('depositBase with one mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, longer, depositer, lender] = await ethers.getSigners();
    await marginlyPool.connect(longer).depositBase(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 0.25;
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(lender).depositQuote(10 * longAmount);
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(depositer).depositBase(1000));
    expect(await marginlyPool.discountedQuoteDebt()).to.be.equal(BigNumber.from(0));
  });

  it('depositQuote with one mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, longer, depositer, lender] = await ethers.getSigners();
    await marginlyPool.connect(longer).depositBase(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 0.25;
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(lender).depositQuote(10 * longAmount);
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(depositer).depositQuote(1000));
    expect(await marginlyPool.discountedQuoteDebt()).to.be.equal(BigNumber.from(0));
  });

  it('short with one mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, depositer, longer, shorter] = await ethers.getSigners();

    await marginlyPool.connect(depositer).depositBase(100);

    await marginlyPool.connect(longer).depositBase(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 0.25;
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(shorter).depositQuote(10 * longAmount);
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(shorter).short(10));
    expect(await marginlyPool.discountedQuoteDebt()).to.be.equal(BigNumber.from(0));
  });

  it('long with one mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, depositer, longer, lender, shorter] = await ethers.getSigners();
    await marginlyPool.connect(depositer).depositQuote(100);
    await marginlyPool.connect(longer).depositBase(1000);
    await marginlyPool.connect(shorter).depositQuote(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 1;
    const price = +toHumanString(BigNumber.from((await marginlyPool.getBasePrice()).inner));
    const shortAmount = Math.floor(((lev - 1) * 1000) / price);
    await marginlyPool.connect(lender).depositBase(3 * shortAmount);
    await marginlyPool.connect(shorter).short(shortAmount);
    await time.increase(24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(longer).long(10));
    expect(await marginlyPool.discountedBaseDebt()).to.be.equal(BigNumber.from(0));
  });

  it('long initialized heap with one mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, longer, lender, longer2] = await ethers.getSigners();
    await marginlyPool.connect(longer).depositBase(1000);
    await marginlyPool.connect(longer2).depositBase(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 0.25;
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(lender).depositQuote(3 * longAmount);
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(longer2).long(10));
    expect(await marginlyPool.discountedBaseDebt()).to.be.equal(BigNumber.from(0));
  });

  it('long closePosition with one mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [owner, longer, lender, longer2] = await ethers.getSigners();
    const params = await marginlyPool.params();
    await marginlyPool.connect(owner).setParameters({ ...params, positionMinAmount: 1 });
    await marginlyPool.connect(longer).depositBase(1000);
    await marginlyPool.connect(longer2).depositBase(1000);
    await marginlyPool.connect(longer2).long(1);
    const lev = (await marginlyPool.params()).maxLeverage - 0.25;
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(lender).depositQuote(3 * longAmount);
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(longer2).closePosition());
    expect(await marginlyPool.discountedBaseDebt()).to.be.equal(BigNumber.from(0));
  });

  it('depositBase with two mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, longer, shorter, lender, depositer] = await ethers.getSigners();
    await marginlyPool.connect(longer).depositBase(1000);
    await marginlyPool.connect(shorter).depositQuote(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 1;
    const price = +toHumanString(BigNumber.from((await marginlyPool.getBasePrice()).inner));
    const shortAmount = Math.floor(((lev - 1) * 1000) / price);
    await marginlyPool.connect(lender).depositBase(3 * shortAmount);
    await marginlyPool.connect(shorter).short(shortAmount);
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(lender).depositQuote(Math.floor(20 * longAmount * price));
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(180 * 24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(depositer).depositBase(1000));
    expect((await marginlyPool.positions(shorter.address)).discountedBaseAmount).to.be.equal(BigNumber.from(0));
    expect((await marginlyPool.positions(longer.address)).discountedQuoteAmount).to.be.equal(BigNumber.from(0));
  });

  it('depositQuote with two mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, longer, shorter, lender, depositer] = await ethers.getSigners();
    await marginlyPool.connect(longer).depositBase(1000);
    await marginlyPool.connect(shorter).depositQuote(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 1;
    const price = +toHumanString(BigNumber.from((await marginlyPool.getBasePrice()).inner));
    const shortAmount = Math.floor(((lev - 1) * 1000) / price);
    await marginlyPool.connect(lender).depositBase(3 * shortAmount);
    await marginlyPool.connect(shorter).short(shortAmount);
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(lender).depositQuote(Math.floor(20 * longAmount * price));
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(180 * 24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(depositer).depositQuote(1000));
    expect((await marginlyPool.positions(shorter.address)).discountedBaseAmount).to.be.equal(BigNumber.from(0));
    expect((await marginlyPool.positions(longer.address)).discountedQuoteAmount).to.be.equal(BigNumber.from(0));
  });

  it('short with two mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, longer, shorter, shorter2, lender, depositer] = await ethers.getSigners();
    await marginlyPool.connect(longer).depositBase(1000);
    await marginlyPool.connect(shorter).depositQuote(1000);
    await marginlyPool.connect(shorter2).depositQuote(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 1;
    const price = +toHumanString(BigNumber.from((await marginlyPool.getBasePrice()).inner));
    const shortAmount = Math.floor(((lev - 1) * 1000) / price);
    await marginlyPool.connect(lender).depositBase(3 * shortAmount);
    await marginlyPool.connect(shorter).short(shortAmount);
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(lender).depositQuote(Math.floor(20 * longAmount * price));
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(180 * 24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(shorter2).short(10));
    expect((await marginlyPool.positions(shorter.address)).discountedBaseAmount).to.be.equal(BigNumber.from(0));
    expect((await marginlyPool.positions(longer.address)).discountedQuoteAmount).to.be.equal(BigNumber.from(0));
  });

  it('long with two mc', async () => {
    const { marginlyPool } = await loadFixture(createMarginlyPool);
    const [_, longer, longer2, shorter, lender, depositer] = await ethers.getSigners();
    await marginlyPool.connect(longer).depositBase(1000);
    await marginlyPool.connect(longer2).depositBase(1000);
    await marginlyPool.connect(shorter).depositQuote(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 1;
    const price = +toHumanString(BigNumber.from((await marginlyPool.getBasePrice()).inner));
    const shortAmount = Math.floor(((lev - 1) * 1000) / price);
    await marginlyPool.connect(lender).depositBase(3 * shortAmount);
    await marginlyPool.connect(shorter).short(shortAmount);
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(lender).depositQuote(Math.floor(20 * longAmount * price));
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(180 * 24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(longer2).long(10));
    expect((await marginlyPool.positions(shorter.address)).discountedBaseAmount).to.be.equal(BigNumber.from(0));
    expect((await marginlyPool.positions(longer.address)).discountedQuoteAmount).to.be.equal(BigNumber.from(0));
  });

  it('closePosition with two mc', async () => {
    const {
      marginlyPool,
      uniswapPoolInfo: { pool },
    } = await loadFixture(createMarginlyPool);
    const [owner, longer, longer2, shorter, lender, depositer] = await ethers.getSigners();
    const params = await marginlyPool.params();
    await marginlyPool.connect(owner).setParameters({ ...params, positionMinAmount: 1 });
    await marginlyPool.connect(longer).depositBase(1000);
    await marginlyPool.connect(longer2).depositBase(1000);
    await marginlyPool.connect(shorter).depositQuote(1000);
    const lev = (await marginlyPool.params()).maxLeverage - 1;
    const price = +toHumanString(BigNumber.from((await marginlyPool.getBasePrice()).inner));
    const shortAmount = Math.floor(((lev - 1) * 1000) / price);
    await marginlyPool.connect(lender).depositBase(3 * shortAmount);
    await marginlyPool.connect(shorter).short(shortAmount);
    const longAmount = Math.floor((lev - 1) * 1000);
    await marginlyPool.connect(lender).depositQuote(Math.floor(20 * longAmount * price));
    await marginlyPool.connect(longer2).long(1);
    await marginlyPool.connect(longer).long(longAmount);
    await time.increase(180 * 24 * 60 * 60);
    await snapshotGasCost(await marginlyPool.connect(longer2).closePosition());
    expect((await marginlyPool.positions(shorter.address)).discountedBaseAmount).to.be.equal(BigNumber.from(0));
    expect((await marginlyPool.positions(longer.address)).discountedQuoteAmount).to.be.equal(BigNumber.from(0));
  });
});

describe('Liquidation', () => {
  it('liquidate long position and create new position', async () => {
    const {
      marginlyPool,
      wallets: [longer, receiver],
    } = await loadFixture(getInitializedPool);

    const baseCollateral = 100;
    await marginlyPool.connect(longer).depositBase(baseCollateral);

    const longAmount = 1970; // leverage 19.8
    await marginlyPool.connect(longer).long(longAmount);

    //wait for accrue interest
    const timeShift = 20 * 24 * 60 * 60;
    await time.increase(timeShift);

    const quoteAmount = 3000;
    const baseAmount = 0;
    await snapshotGasCost(marginlyPool.connect(receiver).receivePosition(longer.address, quoteAmount, baseAmount));
  });

  it('liquidate short position and create new position', async () => {
    const {
      marginlyPool,
      wallets: [shorter, receiver],
    } = await loadFixture(getInitializedPool);

    const shorterCollateral = 100;
    await marginlyPool.connect(shorter).depositQuote(shorterCollateral);

    const shortAmount = 7600; // leverage 19.9
    await marginlyPool.connect(shorter).short(shortAmount);

    //wait for accrue interest
    const timeShift = 20 * 24 * 60 * 60;
    await time.increase(timeShift);

    const quoteAmount = 356;
    const baseAmount = 7700; // the sum is enough to cover debt + accruedInterest
    await snapshotGasCost(marginlyPool.connect(receiver).receivePosition(shorter.address, quoteAmount, baseAmount));
  });

  it('liquidate long position and create new long position', async () => {
    const {
      marginlyPool,
      wallets: [longer, receiver],
    } = await loadFixture(getInitializedPool);

    const baseCollateral = 100;
    await marginlyPool.connect(longer).depositBase(baseCollateral);

    const longAmount = 1970; // leverage 19.8
    await marginlyPool.connect(longer).long(longAmount);

    //wait for accrue interest
    const timeShift = 20 * 24 * 60 * 60;
    await time.increase(timeShift);

    const quoteAmount = 200; // the sum is not enough to cover bad position debt
    const baseAmount = 0;
    await snapshotGasCost(marginlyPool.connect(receiver).receivePosition(longer.address, quoteAmount, baseAmount));
  });

  it('liquidate short position and create new short position', async () => {
    const {
      marginlyPool,
      wallets: [shorter, receiver],
    } = await loadFixture(getInitializedPool);

    const shorterCollateral = 100;
    await marginlyPool.connect(shorter).depositQuote(shorterCollateral);

    const shortAmount = 7600; // leverage 19.9
    await marginlyPool.connect(shorter).short(shortAmount);

    //wait for accrue interest
    const timeShift = 20 * 24 * 60 * 60;
    await time.increase(timeShift);

    const quoteAmount = 1000; // the sum is enough to improove position leverage
    const baseAmount = 100; // the sum is not enough to cover debt + accruedInterest
    await snapshotGasCost(marginlyPool.connect(receiver).receivePosition(shorter.address, quoteAmount, baseAmount));
  });
});
