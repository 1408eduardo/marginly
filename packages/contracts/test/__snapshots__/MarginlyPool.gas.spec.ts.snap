// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deposit into existing position: depositBase 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 136746,
}
`;

exports[`Deposit into existing position: depositQuote 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 136904,
}
`;

exports[`Liquidation liquidate long position and create new long position 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 271546,
}
`;

exports[`Liquidation liquidate long position and create new position 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 299392,
}
`;

exports[`Liquidation liquidate short position and create new position 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 307719,
}
`;

exports[`Liquidation liquidate short position and create new short position 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 282565,
}
`;

exports[`Open position: depositBase 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 218073,
}
`;

exports[`Open position: depositQuote 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 218204,
}
`;

exports[`System initialized: closePosition 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 237126,
}
`;

exports[`System initialized: depositBase 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 216938,
}
`;

exports[`System initialized: depositBase and long 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 367488,
}
`;

exports[`System initialized: depositQuote 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 217023,
}
`;

exports[`System initialized: depositQuote and short 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 346354,
}
`;

exports[`System initialized: long 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 308261,
}
`;

exports[`System initialized: short 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 286980,
}
`;

exports[`System initialized: withdrawBase 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 186480,
}
`;

exports[`System initialized: withdrawQuote 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 194897,
}
`;

exports[`mc happens: MC long position with deleverage 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 226674,
}
`;

exports[`mc happens: MC long reinit 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 195675,
}
`;

exports[`mc happens: MC short position with deleverage 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 226742,
}
`;

exports[`mc happens: MC short reinit 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 195628,
}
`;

exports[`mc happens: closePosition with two mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 375084,
}
`;

exports[`mc happens: depositBase with one mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 261373,
}
`;

exports[`mc happens: depositBase with two mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 360197,
}
`;

exports[`mc happens: depositQuote with one mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 252331,
}
`;

exports[`mc happens: depositQuote with two mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 290590,
}
`;

exports[`mc happens: long closePosition with one mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 265182,
}
`;

exports[`mc happens: long initialized heap with one mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 291298,
}
`;

exports[`mc happens: long with one mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 368473,
}
`;

exports[`mc happens: long with two mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 389102,
}
`;

exports[`mc happens: short with one mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 336658,
}
`;

exports[`mc happens: short with two mc 1`] = `
Object {
  "calldataByteLength": 260,
  "gasUsed": 384669,
}
`;
